File: app/routes/movies.py
from flask import Blueprint, request, jsonify, current_app

movies_bp = Blueprint("movies", __name__)

SORT_FIELD_MAP = {
    "release_date": "release_date",
    "rating": "vote_average",
}


@movies_bp.route("/movies", methods=["GET"])
def list_movies():
    db = current_app.db
    collection = db["movies"]

    # --- pagination ---
    try:
        page = max(1, int(request.args.get("page", 1)))
        limit = min(100, max(1, int(request.args.get("limit", 10))))
    except ValueError:
        return jsonify({"error": "Invalid pagination params"}), 400

    # --- filters ---
    query = {}
    year = request.args.get("year")
    if year and year != "all":
        try:
            query["release_year"] = int(year)
        except ValueError:
            return jsonify({"error": "Invalid year"}), 400

    language = request.args.get("language")
    if language and language.lower() != "all":
        query["languages"] = {"$in": [language]}

    # --- sorting ---
    sort_by = request.args.get("sort_by", "release_date")
    order = request.args.get("order", "desc")
    mongo_field = SORT_FIELD_MAP.get(sort_by, "release_date")
    mongo_order = -1 if order == "desc" else 1

    skip = (page - 1) * limit
    total = collection.count_documents(query)

    cursor = (
        collection.find(query, {"_id": 0})
        .sort(mongo_field, mongo_order)
        .skip(skip)
        .limit(limit)
    )

    movies = []
    for doc in cursor:
        doc["release_date"] = (
            doc["release_date"].strftime("%d %b %Y") if doc.get("release_date") else None
        )
        movies.append(doc)

    return jsonify({
        "page": page,
        "limit": limit,
        "total": total,
        "total_pages": (total + limit - 1) // limit,
        "movies": movies,
    }), 200


@movies_bp.route("/movies/filters", methods=["GET"])
def get_filter_options():
    """Return distinct years and languages for filter dropdowns."""
    db = current_app.db
    collection = db["movies"]

    years = sorted(
        [y for y in collection.distinct("release_year") if y is not None],
        reverse=True,
    )
    languages = sorted(
        [l for l in collection.distinct("languages") if l]
    )

    return jsonify({"years": years, "languages": languages}), 200

File: static/js/movies.js
let currentPage = 1;
let currentSortBy = 'release_date';
let currentOrder = 'desc';

// Load filter dropdowns on page load
document.addEventListener('DOMContentLoaded', () => {
  loadFilterOptions();
  loadMovies();
});

function loadFilterOptions() {
  fetch('/api/movies/filters')
    .then(r => r.json())
    .then(data => {
      const yearSel = document.getElementById('filter-year');
      data.years.forEach(y => {
        const opt = document.createElement('option');
        opt.value = y; opt.textContent = y;
        yearSel.appendChild(opt);
      });

      const langSel = document.getElementById('filter-language');
      data.languages.forEach(l => {
        const opt = document.createElement('option');
        opt.value = l; opt.textContent = l;
        langSel.appendChild(opt);
      });
    })
    .catch(() => {});
}

function loadMovies(page = 1) {
  currentPage = page;
  const year = document.getElementById('filter-year').value;
  const language = document.getElementById('filter-language').value;
  const sortBy = document.getElementById('sort-by').value;
  const order = document.getElementById('sort-order').value;
  currentSortBy = sortBy;
  currentOrder = order;

  updateSortIcons();

  const params = new URLSearchParams({
    page, limit: 10, sort_by: sortBy, order,
    ...(year !== 'all' && { year }),
    ...(language !== 'all' && { language }),
  });

  const tbody = document.getElementById('movies-tbody');
  tbody.innerHTML = '<tr><td colspan="6" class="loading-cell">Loading...</td></tr>';

  fetch(`/api/movies?${params}`)
    .then(r => r.json())
    .then(data => {
      renderTable(data, page);
      renderPagination(data);
    })
    .catch(() => {
      tbody.innerHTML = '<tr><td colspan="6" class="empty-cell">Failed to load data.</td></tr>';
    });
}

function renderTable(data, page) {
  const tbody = document.getElementById('movies-tbody');
  if (!data.movies || data.movies.length === 0) {
    tbody.innerHTML = '<tr><td colspan="6" class="empty-cell">No records found.</td></tr>';
    return;
  }

  const offset = (page - 1) * data.limit;
  tbody.innerHTML = data.movies.map((m, i) => {
    const lang = m.languages && m.languages.length > 0 ? m.languages[0] : (m.original_language || '—');
    const rating = m.vote_average != null ? m.vote_average.toFixed(1) : '—';
    return `<tr>
      <td>${offset + i + 1}</td>
      <td>${escHtml(m.title || m.original_title || '—')}</td>
      <td>${m.release_date || '—'}</td>
      <td>${m.release_year || '—'}</td>
      <td>${escHtml(lang)}</td>
      <td>${rating}</td>
    </tr>`;
  }).join('');
}

function renderPagination(data) {
  const info = document.getElementById('pagination-info');
  const controls = document.getElementById('pagination-controls');
  const { page, limit, total, total_pages } = data;

  const start = Math.min((page - 1) * limit + 1, total);
  const end = Math.min(page * limit, total);
  info.textContent = `Showing ${start} to ${end} of ${total.toLocaleString()} records${total < data._total_unfiltered ? ' (filtered)' : ''}`;

  controls.innerHTML = '';

  const prevBtn = pageBtn('←', page <= 1, () => loadMovies(page - 1));
  controls.appendChild(prevBtn);

  const pages = pagesToShow(page, total_pages);
  let lastPage = 0;
  pages.forEach(p => {
    if (p - lastPage > 1) {
      const ell = document.createElement('span');
      ell.className = 'page-ellipsis'; ell.textContent = '...';
      controls.appendChild(ell);
    }
    const btn = pageBtn(p, false, () => loadMovies(p));
    if (p === page) btn.classList.add('active');
    controls.appendChild(btn);
    lastPage = p;
  });

  const nextBtn = pageBtn('→', page >= total_pages, () => loadMovies(page + 1));
  controls.appendChild(nextBtn);
}

function pagesToShow(current, total) {
  const pages = new Set([1, total, current, current - 1, current + 1].filter(p => p >= 1 && p <= total));
  return [...pages].sort((a, b) => a - b);
}

function pageBtn(label, disabled, onClick) {
  const btn = document.createElement('button');
  btn.className = 'page-btn';
  btn.textContent = label;
  btn.disabled = disabled;
  if (!disabled) btn.addEventListener('click', onClick);
  return btn;
}

function applyFilters() { loadMovies(1); }

function resetFilters() {
  document.getElementById('filter-year').value = 'all';
  document.getElementById('filter-language').value = 'all';
  document.getElementById('sort-by').value = 'release_date';
  document.getElementById('sort-order').value = 'desc';
  loadMovies(1);
}

function toggleSort(field) {
  if (currentSortBy === field) {
    document.getElementById('sort-order').value = currentOrder === 'desc' ? 'asc' : 'desc';
  } else {
    document.getElementById('sort-by').value = field;
    document.getElementById('sort-order').value = 'desc';
  }
  loadMovies(1);
}

function updateSortIcons() {
  ['release_date', 'rating'].forEach(f => {
    const el = document.getElementById(`sort-icon-${f}`);
    if (!el) return;
    if (f === currentSortBy) {
      el.textContent = currentOrder === 'asc' ? '↑' : '↓';
    } else {
      el.textContent = '↕';
    }
  });
}

function escHtml(str) {
  return String(str).replace(/&/g,'&amp;').replace(/</g,'&lt;').replace(/>/g,'&gt;');
}
